bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/RuleEngine.java

```diff
@@ -0,0 +1,426 @@
+ * http://www.prosyst.com
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    ProSyst Software GmbH - initial API and implementation
+ *******************************************************************************/
+
+package org.eclipse.smarthome.automation.core;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.smarthome.automation.Action;
+import org.eclipse.smarthome.automation.Condition;
+import org.eclipse.smarthome.automation.Connection;
+import org.eclipse.smarthome.automation.Module;
+import org.eclipse.smarthome.automation.Rule;
+import org.eclipse.smarthome.automation.Trigger;
+import org.eclipse.smarthome.automation.core.RuleEngineCallbackImpl.TriggerData;
+import org.eclipse.smarthome.automation.handler.ActionHandler;
+import org.eclipse.smarthome.automation.handler.ConditionHandler;
+import org.eclipse.smarthome.automation.handler.ModuleHandler;
+import org.eclipse.smarthome.automation.handler.ModuleHandlerFactory;
+import org.eclipse.smarthome.automation.handler.RuleEngineCallback;
+import org.eclipse.smarthome.automation.handler.TriggerHandler;
+import org.eclipse.smarthome.automation.type.Input;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceReference;
+import org.osgi.util.tracker.ServiceTracker;
+import org.osgi.util.tracker.ServiceTrackerCustomizer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author Yordan Mihaylov - Initial Contribution
+ */
+public class RuleEngine implements ServiceTrackerCustomizer/* <ModuleHandlerFactory, ModuleHandlerFactory> */{
+
+    public static final int MODULE_TYPE_SEPARATOR = ':';
+
+    private static Map<String, RuleImpl> rules;
+    private static Map<String, Thread> runningRules;
+    private ServiceTracker/* <ModuleHandlerFactory, ModuleHandlerFactory> */msfTracker;
+    private BundleContext bc;
+    private Map<String, ModuleHandlerFactory> moduleHandlerFactories;
+    private Logger log;
+
+    public RuleEngine(BundleContext bc) {
+        this.bc = bc;
+        this.log = LoggerFactory.getLogger(getClass());
+        if (rules == null) {
+            rules = new HashMap<String, RuleImpl>(20);
+        }
+        if (runningRules == null) {
+            runningRules = new HashMap<String, Thread>(20);
+        }
+        moduleHandlerFactories = new HashMap<String, ModuleHandlerFactory>(20);
+        msfTracker = new ServiceTracker/* <ModuleHandlerFactory, ModuleHandlerFactory> */(bc,
+                ModuleHandlerFactory.class.getName(), this);
+        msfTracker.open();
+
+    }
+
+    protected void setRule(RuleImpl r) {
```

Comment:
2) Good question. It is not really content OF the rule, but rather ABOUT it, so I would not want to add it to the rule interface. From a REST perspective, I would possibly expect it to be available under /rules/xyz/status. So it is probably the RuleManager which should offer some additional methods to return a RuleStatus (or a List to return all from a list of rules).
4) You mean the RuleProvider? It should return all rules as the provider itself actually cannot know whether all handlers are available or not. Only the RuleManager knows.
5) This is the comment I referred to regarding handler tracking: https://github.com/eclipse/smarthome/pull/317/files#r32005473